python-sqlite-sqlalchemy/project/examples/example_2/main.py

```diff
@@ -1,77 +1,185 @@
+    first_name, last_name = author_name.split(" ")
+    author = (
+        session.query(Author)
+        .filter(
+            and_(
+                Author.first_name == first_name, Author.last_name == last_name
+            )
+        )
+        .one_or_none()
+    )
+
+    # Get the book if exists
+    book = session.query(Book).filter(Book.title == book_title).one_or_none()
+
+    # Get the publisher if exists
+    publisher = (
+        session.query(Publisher)
+        .filter(Publisher.name == publisher_name)
+        .one_or_none()
+    )
+    # Does book not exist, and the author or publisher also not existexist?
+    if book is not None and not (author is None or publisher is None):
+        raise Exception(
+            "New item exists", author_name, book_title, publisher_name
+        )
+    # Create the book
+    book = Book(title=book_title)
+
+    # Add the book to the author's books collection if didn't exist
+    author.books.append(book)
+
+    # Add the book to the publisher's collection if didn't exist
+    publisher.books.append(book)
+
+    # Commit to the database
+    session.commit()
+
+
+def output_author_hierarchy(authors):
+    """
+    Outputs the author/book/publisher information in
+    a hierarchical manner
+
+    :param authors:         the collection of root author objects
+    :return:                None
+    """
+    authors_tree = Tree()
+    authors_tree.create_node("Authors", "authors")
+    for author in authors:
+        authors_tree.create_node(
+            f"{author.first_name} {author.last_name}",
+            f"{author.first_name} {author.last_name}",
+            parent="authors",
+        )
+        for book in author.books:
+            authors_tree.create_node(
+                f"{book.title}",
+                f"{book.title}",
+                parent=f"{author.first_name} {author.last_name}",
+            )
+            for publisher in book.publishers:
+                authors_tree.create_node(
+                    f"{publisher.name}", uuid4(), parent=f"{book.title}"
+                )
+    # Output the hierarchical authors data
+    authors_tree.show()
 
-    # Connect to the sqlite database
-    sqlite_filepath = resource_filename("project.data", "temp_data.db")
-    connection = sqlite3.connect(sqlite_filepath)
 
-    # Get the average temperature by date
-    date_string = "2019-02-10"
-    average_temp = get_average_temp_by_date(date_string, connection)
-    print(f"Average temp {date_string}: {average_temp:.2f}")
+def main():
+    """Main entry point of program"""
+
+    # Connect to the database using SqlAlchemy
+    sqlite_filepath = resource_filename(
+        "project.data", "author_book_publisher.db"
+    )
+    engine = create_engine(f"sqlite:///{sqlite_filepath}")
+    Session = sessionmaker()
+    Session.configure(bind=engine)
+    session = Session()
+
+    # Get the total number of books printed by each publisher
+    books_by_publisher = get_books_by_publishers(session)
```

Comment:
To be consistent with Example 1, we should also change the direction:

```suggestion
    books_by_publisher = get_books_by_publishers(session, ascending=False)
```

Similarly, when getting `authors_by_publisher`.